### DISCUSSION TITLE
Adding files to zip fails when using AES

### DISCUSSION THREAD

**Discussion by fedepr** [icsharpcode/SharpZipLib#506] [General] [answered]
### Steps to reproduce
Hi everyone, i've got this code
```cs
public static void CreateZipPackage(string zipName, Dictionary<string, byte[]> files)
{
    using(FileStream zip = File.Create(zipName))
    {
        using(ZipOutputStream zipStream = new ZipOutputStream(zip))
        {
            foreach (KeyValuePair<string, byte[]> kv in files)
            {

                // Compress the entry in the package
                ZipEntry entry = new ZipEntry(kv.Key);
                entry.AESKeySize = 256;
                //Add an entry
                zipStream.PutNextEntry(entry);
                // Set the compression level 1~9
                zipStream.SetLevel(5);
                zipStream.Password = "PASSWORD";
                //write
                zipStream.Write(kv.Value, 0, kv.Value.Length);

            }
        }
    }
}
```
I noticed that i can view the content of the files in the ZIP even if it's password protected, so i've decided to add "entry.AESKeySize = 256;" to encrypt the files
But then i received the Error: Buffer cannot be null - parameter name: array (when closing the ZipOutputStream)

### Expected behavior
I want to zip and encrypt multiple files in c#

### Actual behavior
Error: Buffer cannot be null - parameter name: array

### Version of SharpZipLib
1.2.0.246

**Comment by piksel** [comment, marked as answer]
Hm, you are setting the password after the entry has been added, that probably doesn't work. Try moving the
```cs
                // Set the compression level 1~9
                zipStream.SetLevel(5);
                zipStream.Password = "PASSWORD";
```
to before the `foreach` loop. 

Also, are you sure about that error message? It doesn't seem right.
If you can give me the full exception I can look into what is going on...

**Comment by Numpsy** [comment]
I tried turning off the password in one of the zip output stream unit tests and got

```
  Message: 
    System.ArgumentNullException : Buffer cannot be null.
    Parameter name: buffer
  Stack Trace: 
    MemoryStream.Write(Byte[] buffer, Int32 offset, Int32 count)
    ZipOutputStream.CloseEntry() line 534
    ZipOutputStream.Finish() line 772
    DeflaterOutputStream.Dispose(Boolean disposing) line 404
    Stream.Close()
    ZipEncryptionHandling.WriteEncryptedZipToStream(Stream stream, String password, Int32 keySize, CompressionMethod compressionMethod) line 497
    ZipEncryptionHandling.CreateZipWithEncryptedEntries(String password, Int32 keySize, CompressionMethod compressionMethod) line 540
```

Because ```AESAuthCode``` is null.

Looks to be because some of the logic goes off entry.AESKeySize being set to a non-zero value (deciding whether to write the AES extra data etc), but them some of it goes off Entry.IsCrypted being true, which is only done automatically if a password has already been specified when the entry is added, so some of the other crypto bits get skipped.

Gets itself in a bit of a mess basically. 
Possibly ```PutNextEntry``` should throw if you try to add an entry with AESKeySize set, but no password set?

**Comment by piksel** [comment]
Yeah, that's basically what I figured was going on. It should definitely throw there, the error messages of the type `x cannot be null` (where x is not from a public parameter) is not particularly helpful for consumers.

I think the original problem was caused by the same problem
> I noticed that i can view the content of the files in the ZIP even if it's password protected
non-AES encryption should still work (even though it's weak and I'd recommend you use AES when possible), but you need to set the password **before** adding the entries.

**Reply by fedepr** [reply]
Hello,
i have tried to move the piece of code of the password setting, now my code is

```cs
public static void CreateZipPackage(string zipName, Dictionary<string, byte[]> files)
{
    using(FileStream zip = File.Create(zipName))
    {
        using(ZipOutputStream zipStream = new ZipOutputStream(zip))
        {
            // Set the compression level 1~9
            zipStream.SetLevel(5);
            zipStream.Password = "PASSWORD";

            foreach (KeyValuePair<string, byte[]> kv in files)
            {

                // Compress the entry in the package
                ZipEntry entry = new ZipEntry(kv.Key);
                entry.AESKeySize = 256;
                //Add an entry
                zipStream.PutNextEntry(entry);

                //write
                zipStream.Write(kv.Value, 0, kv.Value.Length);

            }
        }
    }
}
```
All is working fine thank you very much
